### PR TITLE
Improve inference

### PR DESCRIPTION
**Pull request by tsdlovell**
try infer_struct if infer_map fails
for dataframe, report all inference errors at once

### CODE REVIEW

**Comment by cpcloud** [comment]
This needs some tests to exercises new functionality and any edge cases this might introduce.

File: ibis/expr/datatypes.py [hunk 1]
```diff
@@ -1070,12 +1070,15 @@ def infer_struct(value):
...
 def infer_map(value):
-    if not value:
-        return Map(null, null)
-    return Map(
-        highest_precedence(map(infer, value.keys())),
-        highest_precedence(map(infer, value.values())),
-    )
+    try:
+        if not value:
+            return Map(null, null)
```
Comment (cpcloud) [inline ibis/expr/datatypes.py]: Can you move this branch outside of the `try:`?

File: ibis/expr/datatypes.py [hunk 2]
```diff
@@ -1232,6 +1235,17 @@ def can_cast_arrays(source, target, **kwargs):
     return castable(source.value_type, target.value_type)
 
 
+@castable.register(Struct, Struct)
+def can_cast_struct(source, target, **kwargs):
+    return (
+        (source.names <= target.names) and
```
Comment (cpcloud) [inline ibis/expr/datatypes.py]: Don't need parens here.

Comment (cpcloud) [reply ibis/expr/datatypes.py]: Is `names` a set?

File: ibis/expr/schema.py [hunk 3]
```diff
@@ -188,22 +188,80 @@ def schema_from_dict(d):
...
+        series = series.dropna()
+        if not len(series):
+            return [dt.null]
+
+    if series.dtype != np.object_:
+        ibis_dtype = dt.dtype(series.dtype)
+    else:
+        series_nona = series.dropna()
+        pandas_dtype = infer_pandas_dtype(series_nona)
+        if pandas_dtype == 'string' or pandas_dtype == 'unicode':
```
Comment (cpcloud) [inline ibis/expr/schema.py]: Can you make this into a dictionary lookup?

File: ibis/expr/schema.py [hunk 4]
```diff
@@ -188,22 +188,80 @@ def schema_from_dict(d):
...
+            msg += '\n'
+        if multi_cols:
+            postfix = '\n'.join(
+                '{}:\n\t{}'.format(col, '\n\t'.join(map(str, typs)))
+                for (col, typs) in multi_cols
+            )
+            msg += (
+                'Multiple types found for columns(s):\n' + postfix
+            )
+        raise TypeError(msg)
```
Comment (cpcloud) [inline ibis/expr/schema.py]: Can you move the `msg` creation into a separate function?

File: ibis/expr/datatypes.py (hunk 2 above)
Comment (cpcloud) [reply ibis/expr/datatypes.py]: Are the semantics: "`source` can be cast to `target` if the names of `source` are a subset or equal to the names of `target` and types corresponding to each of those names in each struct type are equivalent"?

Comment (tsdlovell) [reply ibis/expr/datatypes.py]: names is of type `odict_keys`
re semantics: yes; re subset: perhaps there should be a check on the nullability of the missing fields?

Comment (cpcloud) [reply ibis/expr/datatypes.py]: I don't think this will work in python 2 because keys are a list IIRC

**Comment by cpcloud** [comment]
Can you rebase? Looks like a random error from BigQuery.

**Comment by cpcloud** [comment]
I'm picking this up

**Comment by xmnlab** [comment]
hey @tsdlovell @cpcloud what is the status of this PR?

do you need any help with this?

**Comment by datapythonista** [comment]
Closing this as stale. @tsdlovell feel free to ping, if you want to finish this, and I'll reopen the PR.
